packages/components/src/inline-edit/hook.js

```diff
@@ -0,0 +1,134 @@
+/**
+ * External dependencies
+ */
+import { isUndefined, negate, noop, flow } from 'lodash';
+/**
+ * WordPress dependencies
+ */
+import { useEffect, useState, useRef } from '@wordpress/element';
+
+/**
+ * @param {import("react").ChangeEvent<HTMLInputElement>} event
+ */
+const cancelEvent = ( event ) => (
+	event.preventDefault(), event.stopPropagation(), event
+);
+
+/**
+ * @param {import("react").ChangeEvent<HTMLInputElement>} event
+ */
+const getEventValue = ( { target: { value } } ) => value;
+
+/**
+ * @param {Function[]} handlers
+ */
+const mergeEvent = ( ...handlers ) =>
+	/**
+	 * @param {import("react").ChangeEvent<HTMLInputElement>} event
+	 */
+	( event ) => handlers.forEach( ( handler = noop ) => handler( event ) );
+
+/**
+ * @typedef Props
+ * @property {(value: string) => boolean} validate predicate
+ * @property {( value: string ) => void} onWrongInput called when a validate predicate fails
+ * @property {( value: string ) => void} onCommit called on enter/blur
+ * @property {string} value input value
+ */
+
+/**
+ * @param {Props} props
+ */
+export default function useInlineEdit( {
+	validate = negate( isUndefined ),
+	onWrongInput = noop,
+	onCommit = noop,
+	value: propValue,
+} ) {
+	const [ isInEditMode, setIsInEditMode ] = useState( false );
+	const [ editingValue, setEditingValue ] = useState( propValue );
+	/** @type {import('react').RefObject<HTMLInputElement>} */
+	const inputRef = useRef( null );
+	/** @type {import('react').RefObject<HTMLInputElement>} */
+	const toggleRef = useRef( null );
+	const isInvalid = negate( validate );
+	const changeToEditMode = () => setIsInEditMode( true );
+	const changeToToggleMode = () => setIsInEditMode( false );
+
+	useEffect( () => {
+		setEditingValue( propValue );
+		if ( isInvalid( value ) ) onWrongInput( value );
+	}, [ propValue ] );
```

Comment:
`isInvalid` and `onWrongInput` functions are being omitted from the effect dependencies. I understand it would be annoying to include them there since those props could be passed as inline functions and would trigger the effect on every re-render. But I think we should at least add a comment explaining that the effect will get out of sync if the `validate` logic is changed while `value` remains the same. I remember @ItsJonQ worked on something similar.

The local `value` is also omitted from the deps.

Can't we just call `onWrongInput` on the `onChange`/`onBlur` event? It would be much simpler than doing that in an effect.

Reply:
>  I remember @ItsJonQ worked on something similar.

@diegohaz Ah! Are you referring to this `usePropRef` pattern I came up with?
https://github.com/ItsJonQ/g2/blob/main/packages/utils/src/hooks/use-prop-ref.js

```js
// example
const propRefs = usePropRef({ value, step })

const increment = useCallback(() => {
  const { value, step } = propRefs.current
  onChange(value + step)
}, [onChange, propRefs])
```

> call onWrongInput on the onChange/onBlur event?

I haven't tried it, but that seems like that would be similar 🤔

Reply:
@diegohaz As far as I understand what you mean is that we want to make it possible the component will react if a validation (or error handling) function changes. From my perspective, we can do only two things: add those functions as the dependencies and the component will rerender every time, which will lead to poor performance or to skip these dependencies (with a comment as you suggest) as I think it is an edge case.

Solution with `usePropRef` would not apply here as the component would not react for a function change, please see the sandbox:
https://codesandbox.io/s/suspicious-sanne-74lv9?file=/src/App.js

We can use this pattern though, in order to add references to the hooks dependency array and not to make it rerender every time. This will make a code cleaner but the problem will remain the same.